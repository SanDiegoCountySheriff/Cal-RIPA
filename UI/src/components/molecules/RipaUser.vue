<template>
  <div class="ripa-user">
    <v-container>
      <template v-if="!adminEditing">
        <template v-if="admin">
          <v-row no-gutters>
            <v-col cols="12" sm="12">
              <ripa-text-input
                v-model="model.id"
                label="ID"
                :disabled="isRowKeyDisabled"
              >
              </ripa-text-input>
            </v-col>
          </v-row>

          <v-row no-gutters>
            <v-col cols="12" sm="12" md="6">
              <div class="md:tw-mr-4">
                <ripa-text-input
                  v-model="model.firstName"
                  label="First Name"
                  :rules="firstNameRules"
                >
                </ripa-text-input>
              </div>
            </v-col>

            <v-col cols="12" sm="12" md="6">
              <div>
                <ripa-text-input
                  v-model="model.lastName"
                  label="Last Name"
                  :rules="lastNameRules"
                >
                </ripa-text-input>
              </div>
            </v-col>
          </v-row>
        </template>

        <v-row no-gutters>
          <v-col cols="12" sm="12" md="6">
            <div class="md:tw-mr-4">
              <ripa-date-picker
                v-model="model.startDate"
                label="Start Date"
                :rules="startDateRules"
                @input="handleInput"
              ></ripa-date-picker>
            </div>
          </v-col>

          <v-col cols="12" sm="12" md="6">
            <div>
              <ripa-number-input
                v-model="model.yearsExperience"
                label="Years of Experience"
                :rules="yearsExperienceRules"
                @input="handleInput"
              >
              </ripa-number-input>
            </div>
          </v-col>
        </v-row>

        <v-row no-gutters>
          <v-col cols="12" sm="12" md="6">
            <div class="md:tw-mr-4">
              <ripa-text-input
                v-model="model.agency"
                label="Agency"
                :rules="agencyRules"
                @input="handleInput"
              >
              </ripa-text-input>
            </div>
          </v-col>

          <v-col cols="12" sm="12" md="6">
            <ripa-select
              v-model="model.assignment"
              label="Officer Assignment"
              :items="getAssignmentItems"
              itemText="name"
              itemValue="value"
              :rules="assignmentRules"
              @input="handleInput"
            >
            </ripa-select>
          </v-col>
        </v-row>

        <template
          v-if="
            model.assignment === 10 ||
            model.assignment === 11 ||
            model.assignment === 12
          "
        >
          <v-row no-gutters>
            <v-col cols="12" sm="12">
              <ripa-text-input
                v-model="model.otherType"
                label="Other Type"
                :rules="otherTypeRules"
                @input="handleInput"
              >
              </ripa-text-input>
            </v-col>
          </v-row>
        </template>
      </template>

      <template v-if="adminEditing">
        <v-row no-gutters>
          <v-col cols="12" sm="12" md="6">
            <div class="md:tw-mr-4">
              <ripa-number-input
                v-model="model.yearsExperience"
                label="Years of Experience"
                :rules="yearsExperienceRules"
                @input="handleInput"
              >
              </ripa-number-input>
            </div>
          </v-col>

          <v-col cols="12" sm="12" md="6">
            <ripa-select
              v-model="model.assignment"
              label="Officer Assignment"
              :items="getAssignmentItems"
              itemText="name"
              itemValue="value"
              :rules="assignmentRules"
              @input="handleInput"
            >
            </ripa-select>
          </v-col>
        </v-row>

        <template
          v-if="
            model.assignment === 10 ||
            model.assignment === 11 ||
            model.assignment === 12
          "
        >
          <v-row no-gutters>
            <v-col cols="12" sm="12">
              <ripa-text-input
                v-model="model.otherType"
                label="Other Type"
                :rules="otherTypeRules"
                @input="handleInput"
              >
              </ripa-text-input>
            </v-col>
          </v-row>
        </template>
      </template>

      <template v-if="version === 2">
        <v-row no-gutters>
          <v-col cols="12" sm="12" md="12">
            <ripa-select
              v-model="model.officerRace"
              :items="raceItems"
              :rules="raceRules"
              @input="handleInput"
              label="Officer Race"
              itemValue="name"
              itemText="name"
              clearable
              multiple
            ></ripa-select>
          </v-col>
        </v-row>

        <v-row no-gutters>
          <v-col cols="12" sm="12" md="6">
            <ripa-select
              v-model="model.gender"
              :items="genderItems"
              :rules="genderRules"
              @input="handleInput"
              label="Officer Gender"
              class="md:tw-mr-4"
              itemValue="name"
              itemText="name"
              clearable
            ></ripa-select>
          </v-col>
          <v-col cols="12" sm="12" md="6">
            <ripa-switch
              v-model="model.officerNonBinary"
              :rules="genderRules"
              :max-width="250"
              @input="handleInput"
              label="Nonbinary Person"
            ></ripa-switch>
          </v-col>
        </v-row>
      </template>
    </v-container>
  </div>
</template>

<script>
import RipaDatePicker from '@/components/atoms/RipaDatePicker'
import RipaNumberInput from '@/components/atoms/RipaNumberInput'
import RipaSelect from '@/components/atoms/RipaSelect'
import RipaTextInput from '@/components/atoms/RipaTextInput'
import RipaSwitch from '@/components/atoms/RipaSwitch'
import {
  OFFICER_ASSIGNMENTS,
  OFFICER_ASSIGNMENTS_V2,
  GENDERS_V2,
  RACES_V2,
} from '@/constants/form'
import {
  isValidDate,
  dateNotInFuture,
  formatShortDate,
  formatToIsoCurrentDate,
} from '@/utilities/dates'

export default {
  name: 'ripa-officer',

  components: {
    RipaDatePicker,
    RipaNumberInput,
    RipaSelect,
    RipaTextInput,
    RipaSwitch,
  },

  data() {
    return {
      genderItems: GENDERS_V2,
      raceItems: RACES_V2,
      viewModel: this.value,
    }
  },

  inject: ['version'],

  computed: {
    model: {
      get() {
        return this.viewModel
      },
    },

    getAssignmentItems() {
      return this.version === 2 ? OFFICER_ASSIGNMENTS_V2 : OFFICER_ASSIGNMENTS
    },

    yearsExperienceRules() {
      return [
        v => !!v || 'Years of Experience is required',
        v =>
          (v >= 1 && v <= 50) ||
          'Years of Experience must be between 1 and 50 Years',
      ]
    },

    agencyRules() {
      return [v => !!v || 'An agency is required']
    },

    assignmentRules() {
      return [v => !!v || 'An assignment is required']
    },

    firstNameRules() {
      return [v => !!v || 'A first name is required']
    },

    lastNameRules() {
      return [v => !!v || 'A last name is required']
    },

    raceRules() {
      return [v => v?.length > 0 || 'An officer race is required']
    },

    genderRules() {
      const gender = this.model.gender
      const checked = this.model.officerNonBinary
      const isValid = gender || checked

      return [
        !!isValid || 'If no gender is selected, nonbinary must be selected',
      ]
    },

    startDateRules() {
      const startDate = this.viewModel.startDate
      const isValid = isValidDate(startDate)
      const isNotFuture = dateNotInFuture(startDate)

      return [
        (startDate && startDate.length > 0) || 'Start date is required',
        isValid || 'Start date is not a valid date',
        isNotFuture || 'Start date is in the future',
      ]
    },

    otherTypeRules() {
      const assignment = this.viewModel.assignment
      const otherType = this.viewModel.otherType
      if (assignment !== 10 && assignment !== 11 && assignment !== 12) {
        return []
      }

      return [
        (otherType && otherType.length > 0) || 'Other type is required',
        v => (v || '').length <= 60 || 'Max 60 characters',
        v => (v || '').length >= 5 || 'Min 5 characters',
      ]
    },

    getMaxDate() {
      return formatToIsoCurrentDate()
    },
  },

  methods: {
    handleInput() {
      this.updateStartDateModel()
      this.updateOtherTypeModel()
      this.$emit('input', this.viewModel)
    },

    updateStartDateModel() {
      const startDate = this.viewModel.startDate
      const isValid = isValidDate(startDate)

      if (isValid) {
        this.viewModel.startDate = formatShortDate(startDate)
      }
    },

    updateOtherTypeModel() {
      if (
        this.viewModel.assignment !== 10 &&
        this.viewModel.assignment !== 11 &&
        this.viewModel.assignment !== 12
      ) {
        this.viewModel.otherType = null
      }
    },
  },

  watch: {
    value(newVal) {
      this.viewModel = newVal
    },
  },

  props: {
    value: {
      type: Object,
      required: true,
    },
    loading: {
      type: Boolean,
      default: false,
    },
    admin: {
      type: Boolean,
      default: false,
    },
    adminEditing: {
      type: Boolean,
      default: false,
    },
    isRowKeyDisabled: {
      type: Boolean,
      default: true,
    },
  },
}
</script>
