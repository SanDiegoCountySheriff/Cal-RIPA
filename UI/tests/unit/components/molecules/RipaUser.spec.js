import RipaUser from '@/components/molecules/RipaUser.vue'
import { mount } from '@vue/test-utils'
import Vuetify from 'vuetify'

describe('Ripa User', () => {
  let wrapper
  let vuetify
  let ripaUserV2
  let ripaUser

  beforeEach(() => {
    vuetify = new Vuetify()
    ripaUserV2 = {
      agency: 'SDSD',
      assignment: 10,
      otherType: 'Data Services',
      startDate: '2014-10-10',
      yearsExperience: 7,
      officerRace: ['Asian'],
      officerGender: 'Cisgender man',
      officerNonBinary: false,
    }
    ripaUser = {
      agency: 'SDSD',
      assignment: 10,
      otherType: 'Data Services',
      startDate: '2014-10-10',
      yearsExperience: 7,
    }
  })

  const factory = (propsData, version) => {
    return mount(RipaUser, {
      vuetify,
      propsData: {
        ...propsData,
      },
      provide: {
        version,
      },
    })
  }

  it('should match snapshot', () => {
    wrapper = mount(RipaUser, {
      vuetify,
      propsData: {
        value: {
          agency: 'SDSD',
          assignment: 10,
          otherType: 'Data Services',
          startDate: '2014-10-10',
          yearsExperience: 7,
          race: ['Asian'],
          gender: 'Cisgender man',
          officerNonBinary: false,
        },
      },
      provide: {
        version: 2,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should validate officer race rules', () => {
    wrapper = factory(
      {
        value: ripaUserV2,
      },
      2,
    )

    expect(wrapper.vm.raceRules[0](null)).toEqual('An officer race is required')
    expect(wrapper.vm.raceRules[0](['test'])).toEqual(true)
  })

  it.skip('should validate officer gender rules', () => {
    wrapper = factory(
      {
        value: {
          agency: 'SDSD',
          assignment: 10,
          otherType: 'Data Services',
          startDate: '2014-10-10',
          yearsExperience: 7,
          race: ['Asian'],
          gender: null,
          officerNonBinary: false,
        },
      },
      2,
    )

    expect(wrapper.vm.genderRules).toEqual([
      'If no gender is selected, nonbinary must be selected',
    ])
  })

  it.skip('should display officer race and gender inputs if current date > 01/01/24', () => {
    wrapper = factory(
      {
        value: ripaUserV2,
      },
      2,
    )

    expect(wrapper.html()).toContain('Officer Gender')
    expect(wrapper.html()).toContain('Officer Race')
  })

  it('should display new assignment off duty for v2 stop', () => {
    ripaUserV2.assignment = 11
    wrapper = factory(
      {
        value: ripaUserV2,
      },
      2,
    )
    expect(wrapper.html()).toContain('Off duty and/or working private event')
  })

  it('should display new assignment contracted by another agency for v2 stop', () => {
    ripaUserV2.assignment = 12
    wrapper = factory(
      {
        value: ripaUserV2,
      },
      2,
    )
    expect(wrapper.html()).toContain(
      'Contracted by another law enforcement agency',
    )
  })

  it('should not display new assignment types for v1 stop', () => {
    wrapper = factory(
      {
        value: ripaUser,
      },
      1,
    )
    expect(wrapper.html()).not.toContain(
      'Off duty and/or working private event',
    )
    expect(wrapper.html()).not.toContain(
      'Contracted by another law enforcement agency',
    )
  })

  it.todo('should validate years experience rules')

  it.todo('should validate agency rules')

  it.todo('should validate assignment rules')

  it.todo('should validate first name rules')

  it.todo('should validate last name rules')

  it.todo('should validate start date rules')

  it.todo('should validate other type rules')

  it.todo('should get max date')

  it.todo('should handle input')

  it.todo('should update start date model')

  it.todo('should update other type model')

  it.todo('should watch value')
})
