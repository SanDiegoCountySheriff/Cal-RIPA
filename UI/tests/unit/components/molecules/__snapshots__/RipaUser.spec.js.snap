// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ripa User should match snapshot 1`] = `
<div class="ripa-user">
  <div class="container">
    <!---->
    <div class="row no-gutters">
      <div class="col-sm-12 col-md-6 col-12">
        <div class="md:tw-mr-4">
          <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field">
            <div class="v-input__control">
              <div class="v-input__slot">
                <div class="v-text-field__slot"><label for="input-3" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Start Date</label><input id="input-3" type="date"></div>
                <div class="v-input__append-inner">
                  <div class="v-input__icon v-input__icon--clear"><button type="button" aria-label="clear icon" tabindex="-1" class="v-icon notranslate v-icon--link mdi mdi-close theme--light"></button></div>
                </div>
                <div class="v-input__append-inner">
                  <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-calendar theme--light"></i></div>
                </div>
              </div>
              <div class="v-text-field__details">
                <div class="v-messages theme--light">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="col-sm-12 col-md-6 col-12">
        <div>
          <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field">
            <div class="v-input__control">
              <div class="v-input__slot">
                <div class="v-text-field__slot"><label for="input-9" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Years of Experience</label><input inputmode="numeric" min="0" id="input-9" type="number"></div>
              </div>
              <div class="v-text-field__details">
                <div class="v-messages theme--light">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="row no-gutters">
      <div class="col-sm-12 col-md-6 col-12">
        <div class="md:tw-mr-4">
          <div class="ripa-text-input">
            <!---->
            <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-text-field__slot"><label for="input-13" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Agency</label><input id="input-13" type="text"></div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light">
                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="col-sm-12 col-md-6 col-12">
        <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--solo-flat v-select">
          <div class="v-input__control">
            <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-17" class="v-input__slot">
              <div class="v-select__slot"><label for="input-17" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Officer Assignment</label>
                <div class="v-select__selections">
                  <div class="v-select__selection v-select__selection--comma">Other</div><input required="required" id="input-17" readonly="readonly" type="text" aria-readonly="false" autocomplete="off">
                </div>
                <div class="v-input__append-inner">
                  <div class="v-input__icon v-input__icon--clear"><button type="button" aria-label="clear icon" tabindex="-1" class="v-icon notranslate v-icon--link mdi mdi-close theme--light"></button></div>
                </div>
                <div class="v-input__append-inner">
                  <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                </div><input type="hidden" value="10">
              </div>
              <div class="v-menu">
                <!---->
              </div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light">
                <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="row no-gutters">
      <div class="col-sm-12 col-12">
        <div class="ripa-text-input">
          <!---->
          <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field">
            <div class="v-input__control">
              <div class="v-input__slot">
                <div class="v-text-field__slot"><label for="input-24" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Other Type</label><input id="input-24" type="text"></div>
              </div>
              <div class="v-text-field__details">
                <div class="v-messages theme--light">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <!---->
    <div class="row no-gutters">
      <div class="col-sm-12 col-md-12 col-12">
        <div class="v-input theme--light v-text-field v-text-field--solo-flat v-select v-select--is-multi" clearable="">
          <div class="v-input__control">
            <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-28" class="v-input__slot">
              <div class="v-select__slot"><label for="input-28" class="v-label theme--light" style="left: 0px; position: absolute;">Officer Race</label>
                <div class="v-select__selections"><input required="required" id="input-28" readonly="readonly" type="text" aria-readonly="false" autocomplete="off"></div>
                <div class="v-input__append-inner">
                  <div></div>
                </div>
                <div class="v-input__append-inner">
                  <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                </div><input type="hidden">
              </div>
              <div class="v-menu">
                <!---->
              </div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light">
                <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="row no-gutters">
      <div class="col-sm-12 col-md-6 col-12">
        <div class="v-input md:tw-mr-4 theme--light v-text-field v-text-field--solo-flat v-select" clearable="">
          <div class="v-input__control">
            <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-34" class="v-input__slot">
              <div class="v-select__slot"><label for="input-34" class="v-label theme--light" style="left: 0px; position: absolute;">Officer Gender</label>
                <div class="v-select__selections"><input required="required" id="input-34" readonly="readonly" type="text" aria-readonly="false" autocomplete="off"></div>
                <div class="v-input__append-inner">
                  <div></div>
                </div>
                <div class="v-input__append-inner">
                  <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                </div><input type="hidden" value="Cisgender man">
              </div>
              <div class="v-menu">
                <!---->
              </div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light">
                <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="col-sm-12 col-md-6 col-12">
        <div class="ripa-switch" style="max-width: 250px;">
          <div class="v-input v-input--reverse v-input--expand theme--light v-input--selection-controls v-input--switch">
            <div class="v-input__control">
              <div class="v-input__slot">
                <div class="v-input--selection-controls__input"><input aria-checked="false" id="input-40" role="switch" type="checkbox" aria-disabled="false" value="">
                  <div class="v-input--selection-controls__ripple"></div>
                  <div class="v-input--switch__track theme--light"></div>
                  <div class="v-input--switch__thumb theme--light">
                    <transition-stub name="fab-transition" mode="out-in"></transition-stub>
                  </div>
                </div><label for="input-40" class="v-label theme--light" style="left: 0px; position: relative;">
                  Nonbinary Person
                </label>
              </div>
              <div class="v-messages theme--light">
                <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
